### PR TITLE
Kindsys: Remove defs, Slot->SchemaInterface

### PR DESCRIPTION
**Pull request by sdboyer**
**What is this feature?**

This does two things:

* Remove all use of CUE definitions (e.g. `#Core` -> `Core`) from kindsys.
* Rename the "[slot](https://docs.google.com/document/d/12vYU-DRlS0yYxvXW52cOH7kSIxmJAER_4_mlA0An_QM/edit#heading=h.m040j2ma3lf)" concept in kindsys to "schema interface".

**Why do we need this feature?**

[CUE definitions](https://cuelang.org/docs/references/spec/#definitions-and-hidden-fields) recursively close structs they contain when referenced. Using them within kindsys - where our kind definitions contain schema definitions which are impacted by the recursive closure - creates problems for no reason other than "it feels like these should be definitions."

I don't know why it took me so long to realize this. 💩 💩 I'll be changing it in thema, too.

The rename of "slot" is long overdue. The name was always bad - it's a contract for what can go in a slot, not the slot itself.

There's almost no behavior change with schema interface - only that we back off of restricting composable kind declarations by plugin type (e.g. a panel plugin _could_ define a queries composable kind now), as [agreed upon](https://github.com/grafana/grafana/issues/59001#issuecomment-1352925983) by @grafana/plugins-platform.

**Who is this feature for?**

Cool kids.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/kindsys/schema_interface.cue [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+//
+// On the consumer side, any core or custom kind author can choose to define a
+// standard Thema composition slot in its contained lineage that uses one of
+// these schema interfaces as its meta-schema. The slot specification in Thema
+// answers "how", for that kind.
+//
+// Composable kinds declared by a plugin are parsed and validated by Grafana's
+// plugin system when a plugin is installed. This gives each Grafana instance a
+// set of all known #Composable kinds ("whats"), which can be narrowed into the
+// subsets ("which") that each known #CoreStructured or #CustomStructured can
```
Comment (spinillos) [inline pkg/kindsys/schema_interface.cue:36]:
```suggestion
// subsets ("which") that each known #Core or #Custom can
```

File: pkg/kindsys/schema_interface.cue [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+// consume. These subsets are injected dynamically into the consumers, resulting
+// in the actual schema
+//
+// For example, in the Thema lineage for the dashboard #CoreStructured kind:
```
Comment (spinillos) [inline pkg/kindsys/schema_interface.cue:40]:
```suggestion
// For example, in the Thema lineage for the dashboard #Core kind:
```

File: pkg/kindsys/schema_interface.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+// IsGroup indicates whether the slot specifies a group lineage - one in which
+// each top-level key represents a distinct schema for objects that are expected
+// to exist in the wild, but objects corresponding to the root of the schema are not
+// expected to exist.
+func (s SchemaInterface) IsGroup() bool {
+	return s.group
+}
+
+func FindSlot(name string) (SchemaInterface, error) {
```
Comment (spinillos) [inline pkg/kindsys/schema_interface.go:60]: FindSchema?

File: pkg/plugins/pfs/decl_parser.go [hunk 4]
```diff
@@ -60,7 +60,7 @@ func (psr *declParser) Parse(root fs.FS) ([]*PluginDecl, error) {
 				continue
 			}
 			decls = append(decls, &PluginDecl{
-				Slot:       slot,
+				Slot:       &slot,
```
Comment (spinillos) [inline pkg/plugins/pfs/decl_parser.go:63]: It should be Schema here.

File: pkg/plugins/pfs/pfs.go [hunk 5]
```diff
@@ -218,21 +218,25 @@ func ParsePluginFS(f fs.FS, rt *thema.Runtime) (*Tree, error) {
...
 		for _, s := range allslots {
 			iv := val.LookupPath(cue.ParsePath(s.slot.Name()))
-			lin, err := bindSlotLineage(iv, s.slot, r.meta, rt)
-			if lin != nil {
-				r.slotimpls[s.slot.Name()] = lin
-			}
-			if err != nil {
-				return nil, err
+			if iv.Exists() {
+				lin, err := bindSlotLineage(iv, s.slot, r.meta, rt)
```
Comment (spinillos) [inline pkg/plugins/pfs/pfs.go:222]: Can we change this slot here?

File: pkg/plugins/pfs/pfs.go [hunk 6]
```diff
@@ -218,21 +218,25 @@ func ParsePluginFS(f fs.FS, rt *thema.Runtime) (*Tree, error) {
...
 			}
 		}
 	}
 
 	return tree, nil
 }
 
-func bindSlotLineage(v cue.Value, s *kindsys.Slot, meta plugindef.PluginDef, rt *thema.Runtime, opts ...thema.BindOption) (thema.Lineage, error) {
-	accept, required := s.ForPluginType(string(meta.Type))
+func bindSlotLineage(v cue.Value, s kindsys.SchemaInterface, meta plugindef.PluginDef, rt *thema.Runtime, opts ...thema.BindOption) (thema.Lineage, error) {
```
Comment (spinillos) [inline pkg/plugins/pfs/pfs.go:236]: bindSchemaLineage

File: pkg/plugins/pfs/pfs.go (hunk 5 above)
Comment (sdboyer) [reply pkg/plugins/pfs/pfs.go:222]: this stuff is all about to change in the next PR (most things in pfs are rewritten), which is why i didn't bother in cleaning it up in its current state

File: pkg/plugins/pfs/pfs.go (hunk 6 above)
Comment (sdboyer) [reply pkg/plugins/pfs/pfs.go:236]: `bindComposableKind` in the next PR :)

File: pkg/kindsys/kindcat_composable.cue [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package kindsys
+
+// Composable is a category of structured kind that provides schema elements for
+// composition into CoreStructured and CustomStructured kinds. Grafana plugins
```
Comment (undef1nd) [inline pkg/kindsys/kindcat_composable.cue:4]:
```suggestion
// Composable is a category of kind that provides schema elements for
// composition into Core and Custom kinds. Grafana plugins
```

Comment (sdboyer) [reply pkg/kindsys/kindcat_composable.cue]: good catch, ty

**Review by undef1nd — APPROVED** [review]

**Pull request merged**